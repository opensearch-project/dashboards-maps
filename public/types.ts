/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AppMountParameters,
  CoreStart,
  SavedObjectsClient,
  ToastsStart,
  ScopedHistory,
  MountPoint,
} from '../../../src/core/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';
import { DataPublicPluginSetup, DataPublicPluginStart } from '../../../src/plugins/data/public';
import { RegionMapPluginSetup } from '../../../src/plugins/region_map/public';
import { EmbeddableSetup, EmbeddableStart } from '../../../src/plugins/embeddable/public';
import { VisualizationsSetup } from '../../../src/plugins/visualizations/public';
import { ConfigSchema } from '../common/config';
import { DataSourceManagementPluginSetup } from '../../../src/plugins/data_source_management/public';

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  savedObjects: SavedObjectsClient;
  data: DataPublicPluginStart;
  embeddable: EmbeddableStart;
}

export interface MapServices extends CoreStart {
  setHeaderActionMenu: AppMountParameters['setHeaderActionMenu'];
  appBasePath: AppMountParameters['history'];
  element: AppMountParameters['element'];
  navigation: NavigationPublicPluginStart;
  toastNotifications: ToastsStart;
  history: AppMountParameters['history'];
  data: DataPublicPluginStart;
  application: CoreStart['application'];
  i18n: CoreStart['i18n'];
  savedObjects: CoreStart['savedObjects'];
  overlays: CoreStart['overlays'];
  embeddable: EmbeddableStart;
  scopedHistory: ScopedHistory;
  chrome: CoreStart['chrome'];
  uiSettings: CoreStart['uiSettings'];
  mapConfig: ConfigSchema;
  dataSourceManagement: DataSourceManagementPluginSetup;
  setActionMenu: (menuMount: MountPoint | undefined) => void;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface CustomImportMapPluginSetup {}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface CustomImportMapPluginStart {}

export interface AppPluginSetupDependencies {
  regionMap: RegionMapPluginSetup;
  embeddable: EmbeddableSetup;
  visualizations: VisualizationsSetup;
  data: DataPublicPluginSetup;
  dataSourceManagement: DataSourceManagementPluginSetup;
}
