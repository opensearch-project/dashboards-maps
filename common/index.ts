/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fromMBtoBytes } from './util';
import {
  ALLOWED_FILE_EXTENSIONS,
  MAX_FILE_PAYLOAD_SIZE,
  MAX_FILE_PAYLOAD_SIZE_IN_MB,
  PLUGIN_ID,
  PLUGIN_NAVIGATION_BAR_ID,
  PLUGIN_NAME,
} from './constants/shared';

export {
  fromMBtoBytes,
  ALLOWED_FILE_EXTENSIONS,
  MAX_FILE_PAYLOAD_SIZE,
  MAX_FILE_PAYLOAD_SIZE_IN_MB,
  PLUGIN_ID,
  PLUGIN_NAVIGATION_BAR_ID,
  PLUGIN_NAME,
};

export const MAP_VECTOR_TILE_BASIC_STYLE = 'https://tiles.maps.opensearch.org/styles/basic.json';
export const MAP_GLYPHS = 'https://tiles.maps.opensearch.org/fonts/{fontstack}/{range}.pbf';
export const MAP_VECTOR_TILE_DATA_SOURCE = 'https://tiles.maps.opensearch.org/data/v1.json';
export const MAP_DEFAULT_MIN_ZOOM = 0;
export const MAP_DEFAULT_MAX_ZOOM = 22;
export const MAP_REFERENCE_LAYER_DEFAULT_OPACITY = 100;
export const MAP_DATA_LAYER_DEFAULT_OPACITY = 70;
export const MAP_LAYER_DEFAULT_MIN_OPACITY = 0;
export const MAP_LAYER_DEFAULT_MAX_OPACITY = 100;
export const MAP_LAYER_DEFAULT_OPACITY_STEP = 1;
export const MAP_LAYER_DEFAULT_BORDER_THICKNESS = 1;
export const DOCUMENTS_DEFAULT_REQUEST_NUMBER = 1000;
export const DOCUMENTS_DEFAULT_SHOW_TOOLTIPS: boolean = false;
export const DOCUMENTS_DEFAULT_TOOLTIPS: string[] = [];
export const DOCUMENTS_DEFAULT_MARKER_SIZE = 5;
export const LAYER_PANEL_SHOW_LAYER_ICON = 'eye';
export const LAYER_PANEL_HIDE_LAYER_ICON = 'eyeClosed';
export const MAX_LAYER_NAME_LIMIT = 35;
export const MAP_LAYER_DEFAULT_NAME = 'Default map';
export const NEW_MAP_LAYER_DEFAULT_PREFIX = 'New layer';

// Starting position [lng, lat] and zoom
export const MAP_INITIAL_STATE = {
  lng: 0,
  lat: 0,
  zoom: 1,
};

export const APP_PATH = {
  LANDING_PAGE_PATH: '/',
  CREATE_MAP: '/create',
  EDIT_MAP: '/:id',
};

export enum DASHBOARDS_MAPS_LAYER_NAME {
  OPENSEARCH_MAP = 'OpenSearch map',
  DOCUMENTS = 'Documents',
  CUSTOM_MAP = 'Custom map',
}

export enum DASHBOARDS_MAPS_LAYER_TYPE {
  OPENSEARCH_MAP = 'opensearch_vector_tile_map',
  DOCUMENTS = 'documents',
  CUSTOM_MAP = 'custom_map',
}

export enum DASHBOARDS_MAPS_LAYER_ICON {
  OPENSEARCH_MAP = 'globe',
  DOCUMENTS = 'document',
  CUSTOM_MAP = 'globe',
}

export enum DASHBOARDS_MAPS_LAYER_DESCRIPTION {
  OPENSEARCH_MAP = 'Use default OpenSearch basemaps.',
  DOCUMENTS = 'View points, lines, and polygons on the map.',
  CUSTOM_MAP = 'Configure maps to use a custom map source.',
}

export const DOCUMENTS = {
  name: DASHBOARDS_MAPS_LAYER_NAME.DOCUMENTS,
  type: DASHBOARDS_MAPS_LAYER_TYPE.DOCUMENTS,
  icon: DASHBOARDS_MAPS_LAYER_ICON.DOCUMENTS,
  description: DASHBOARDS_MAPS_LAYER_DESCRIPTION.DOCUMENTS,
};

export const OPENSEARCH_MAP_LAYER = {
  name: DASHBOARDS_MAPS_LAYER_NAME.OPENSEARCH_MAP,
  type: DASHBOARDS_MAPS_LAYER_TYPE.OPENSEARCH_MAP,
  icon: DASHBOARDS_MAPS_LAYER_ICON.OPENSEARCH_MAP,
  description: DASHBOARDS_MAPS_LAYER_DESCRIPTION.OPENSEARCH_MAP,
};

export const CUSTOM_MAP = {
  name: DASHBOARDS_MAPS_LAYER_NAME.CUSTOM_MAP,
  type: DASHBOARDS_MAPS_LAYER_TYPE.CUSTOM_MAP,
  icon: DASHBOARDS_MAPS_LAYER_ICON.CUSTOM_MAP,
  description: DASHBOARDS_MAPS_LAYER_DESCRIPTION.CUSTOM_MAP,
};

export interface Layer {
  name: DASHBOARDS_MAPS_LAYER_NAME;
  type: DASHBOARDS_MAPS_LAYER_TYPE;
  icon: DASHBOARDS_MAPS_LAYER_ICON;
  description: DASHBOARDS_MAPS_LAYER_DESCRIPTION;
}

export const LAYER_VISIBILITY = {
  NONE: 'none',
  VISIBLE: 'visible',
};

export const LAYER_ICON_TYPE_MAP: { [key: string]: string } = {
  [DASHBOARDS_MAPS_LAYER_TYPE.OPENSEARCH_MAP]: 'globe',
  [DASHBOARDS_MAPS_LAYER_TYPE.DOCUMENTS]: 'document',
  [DASHBOARDS_MAPS_LAYER_TYPE.CUSTOM_MAP]: 'globe',
};
