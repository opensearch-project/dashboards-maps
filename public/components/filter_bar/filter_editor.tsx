/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiCodeEditor,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiPopoverTitle,
  EuiSpacer,
} from '@elastic/eui';
import React, { useState } from 'react';
import { i18n } from '@osd/i18n';

interface Props {
  content: string;
  label: string;
  onSubmit: (content: string, label: string) => void;
  onCancel: () => void;
}

const isFilterValid = (content: string) => {
  try {
    return Boolean(JSON.parse(content));
  } catch (e) {
    return false;
  }
};

export const FilterEditor = ({ content, label, onSubmit, onCancel }: Props) => {
  const [filterLabel, setFilterLabel] = useState<string>(label);
  const [filterContent, setFilterContent] = useState<string>(content);

  const renderEditor = () => {
    return (
      <EuiFormRow
        label={i18n.translate('maps.filter.filterEditor.parameters', {
          defaultMessage: 'Spatial filter parameters',
        })}
        fullWidth={true}
      >
        <EuiCodeEditor
          value={filterContent}
          onChange={setFilterContent}
          mode="json"
          width="100%"
          height="250px"
        />
      </EuiFormRow>
    );
  };
  return (
    <div>
      <EuiPopoverTitle>
        <EuiFlexGroup alignItems="baseline" responsive={false}>
          <EuiFlexItem>{'Edit Filter'}</EuiFlexItem>
        </EuiFlexGroup>
      </EuiPopoverTitle>

      <div className="globalFilterItem__editorForm">
        <EuiForm>
          {renderEditor()}
          <div>
            <EuiSpacer size="m" />
            <EuiFormRow
              fullWidth={true}
              label={i18n.translate('maps.filter.filterEditor.createCustomLabelInputLabel', {
                defaultMessage: 'Custom label',
              })}
            >
              <EuiFieldText
                fullWidth={true}
                value={filterLabel}
                onChange={(event) => setFilterLabel(event.target.value)}
              />
            </EuiFormRow>
          </div>
          <EuiSpacer size="m" />
          <EuiFlexGroup direction="rowReverse" alignItems="center" responsive={false}>
            <EuiFlexItem grow={false}>
              <EuiButton
                fill
                onClick={() => onSubmit(filterContent, filterLabel)}
                isDisabled={!isFilterValid(filterContent)}
                data-test-subj="saveFilter"
              >
                {'Save'}
              </EuiButton>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty flush="right" onClick={onCancel} data-test-subj="cancelSaveFilter">
                {'Cancel'}
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem />
          </EuiFlexGroup>
        </EuiForm>
      </div>
    </div>
  );
};
