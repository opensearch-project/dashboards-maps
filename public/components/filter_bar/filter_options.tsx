/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButtonIcon, EuiContextMenu, EuiPopover, EuiPopoverTitle } from '@elastic/eui';
import React, { useState } from 'react';
import { i18n } from '@osd/i18n';

interface Props {
  onEnableAll: () => void;
  onDisableAll: () => void;
  onToggleAllNegated: () => void;
  onToggleAllDisabled: () => void;
  onRemoveAll: () => void;
}

export const FilterOptions = ({
  onEnableAll,
  onDisableAll,
  onToggleAllNegated,
  onToggleAllDisabled,
  onRemoveAll,
}: Props) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState<boolean>(false);

  const togglePopover = () => {
    setIsPopoverOpen(!isPopoverOpen);
  };
  const closePopover = () => {
    setIsPopoverOpen(false);
  };

  const panelTree = {
    id: 0,
    items: [
      {
        name: i18n.translate('maps.filter.options.enableAllFiltersButtonLabel', {
          defaultMessage: 'Enable all',
        }),
        icon: 'eye',
        onClick: () => {
          closePopover();
          onEnableAll();
        },
        'data-test-subj': 'enableAllFilters',
      },
      {
        name: i18n.translate('maps.filter.options.disableAllFiltersButtonLabel', {
          defaultMessage: 'Disable all',
        }),
        icon: 'eyeClosed',
        onClick: () => {
          closePopover();
          onDisableAll();
        },
        'data-test-subj': 'disableAllFilters',
      },
      {
        name: i18n.translate('maps.filter.options.invertNegatedFiltersButtonLabel', {
          defaultMessage: 'Invert inclusion',
        }),
        icon: 'invert',
        onClick: () => {
          closePopover();
          onToggleAllNegated();
        },
        'data-test-subj': 'invertInclusionAllFilters',
      },
      {
        name: i18n.translate('maps.filter.options.invertDisabledFiltersButtonLabel', {
          defaultMessage: 'Invert enabled/disabled',
        }),
        icon: 'eye',
        onClick: () => {
          closePopover();
          onToggleAllDisabled();
        },
        'data-test-subj': 'invertEnableDisableAllFilters',
      },
      {
        name: i18n.translate('maps.filter.options.deleteAllFiltersButtonLabel', {
          defaultMessage: 'Remove all',
        }),
        icon: 'trash',
        onClick: () => {
          closePopover();
          onRemoveAll();
        },
        'data-test-subj': 'removeAllFilters',
      },
    ],
  };

  return (
    <EuiPopover
      id="popoverForAllFilters"
      className="globalFilterGroup__allFiltersPopover"
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      button={
        <EuiButtonIcon
          onClick={togglePopover}
          iconType="filter"
          aria-label={i18n.translate('maps.filter.options.changeAllFiltersButtonLabel', {
            defaultMessage: 'Change all filters',
          })}
          title={i18n.translate('maps.filter.options.changeAllFiltersButtonLabel', {
            defaultMessage: 'Change all filters',
          })}
          data-test-subj="showFilterActions"
        />
      }
      anchorPosition="rightUp"
      panelPaddingSize="none"
      repositionOnScroll
    >
      <EuiPopoverTitle>
        {i18n.translate('maps.filter.changeAllFiltersTitle', {
          defaultMessage: 'Change all filters',
        })}
      </EuiPopoverTitle>
      <EuiContextMenu initialPanelId={0} panels={[panelTree]} />
    </EuiPopover>
  );
};
