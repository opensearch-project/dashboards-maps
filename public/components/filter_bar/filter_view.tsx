/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiBadge } from '@elastic/eui';
import { i18n } from '@osd/i18n';
import React, { FC } from 'react';

interface Props {
  isDisabled: boolean;
  valueLabel: string;
  onRemove: () => void;
  [propName: string]: any;
}

export const FilterView: FC<Props> = ({
  isDisabled,
  onRemove,
  onClick,
  valueLabel,
  ...rest
}: Props) => {
  let title = i18n.translate('maps.filter.filterBar.moreFilterActionsMessage', {
    defaultMessage: 'Filter: {valueLabel}. Select for more filter actions.',
    values: { valueLabel },
  });

  if (isDisabled) {
    title = `${i18n.translate('maps.filter.filterBar.disabledFilterPrefix', {
      defaultMessage: 'Disabled',
    })} ${title}`;
  }

  return (
    <EuiBadge
      title={title}
      color="hollow"
      iconType="cross"
      iconSide="right"
      closeButtonProps={{
        // Removing tab focus on close button because the same option can be obtained through the context menu
        // Also, we may want to add a `DEL` keyboard press functionality
        tabIndex: -1,
      }}
      iconOnClick={onRemove}
      iconOnClickAriaLabel={i18n.translate('maps.filter.filterBar.filterItemBadgeIconAriaLabel', {
        defaultMessage: 'Delete',
      })}
      onClick={onClick}
      onClickAriaLabel={i18n.translate('maps.filter.filterBar.filterItemBadgeAriaLabel', {
        defaultMessage: 'Filter actions',
      })}
      {...rest}
    >
      {valueLabel}
    </EuiBadge>
  );
};
