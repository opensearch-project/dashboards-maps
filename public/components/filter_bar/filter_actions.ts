/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { GeoShapeFilterMeta } from '../../../../../src/plugins/data/common';

export const enableGeoShapeFilterMeta = (meta: GeoShapeFilterMeta) =>
  !meta.disabled ? meta : toggleGeoShapeFilterMetaDisabled(meta);

export const disableGeoShapeFilterMeta = (meta: GeoShapeFilterMeta) =>
  meta.disabled ? meta : toggleGeoShapeFilterMetaDisabled(meta);

export const toggleGeoShapeFilterMetaNegated = (meta: GeoShapeFilterMeta) => {
  const negate = !meta.negate;
  return { ...meta, negate };
};
export const toggleGeoShapeFilterMetaDisabled = (meta: GeoShapeFilterMeta) => {
  const status: boolean = !meta.disabled;
  return { ...meta, disabled: status };
};
