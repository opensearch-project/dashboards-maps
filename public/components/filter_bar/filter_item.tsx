/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiContextMenu, EuiPopover } from '@elastic/eui';
import classNames from 'classnames';
import React, { MouseEvent, useState } from 'react';
import { i18n } from '@osd/i18n';
import { GeoShapeFilterMeta } from '../../../../../src/plugins/data/common';
import { FilterView } from './filter_view';
import {
  toggleGeoShapeFilterMetaDisabled,
  toggleGeoShapeFilterMetaNegated,
} from './filter_actions';
import { FilterEditor } from './filter_editor';

interface Props {
  id: string;
  filterMeta: GeoShapeFilterMeta;
  className?: string;
  onUpdate: (filter: GeoShapeFilterMeta) => void;
  onRemove: () => void;
}

export function FilterItem({ filterMeta, onUpdate, onRemove, id, className }: Props) {
  const [isPopoverOpen, setIsPopoverOpen] = useState<boolean>(false);

  function onSubmit(content: string, label: string) {
    setIsPopoverOpen(false);
    const updatedFilter: GeoShapeFilterMeta = {
      ...filterMeta,
      params: JSON.parse(content),
      alias: label,
    };
    onUpdate(updatedFilter);
  }

  function handleClick(e: MouseEvent<HTMLInputElement>) {
    if (e.shiftKey) {
      onToggleDisabled();
    } else {
      setIsPopoverOpen(!isPopoverOpen);
    }
  }

  function onToggleNegated() {
    const f = toggleGeoShapeFilterMetaNegated(filterMeta);
    onUpdate(f);
  }

  function onToggleDisabled() {
    const f = toggleGeoShapeFilterMetaDisabled(filterMeta);
    onUpdate(f);
  }

  function getClasses(negate: boolean, disabled: boolean) {
    return classNames(
      'globalFilterItem',
      {
        'globalFilterItem-isDisabled': disabled,
        'globalFilterItem-isExcluded': negate,
      },
      className
    );
  }

  function getDataTestSubj() {
    const dataTestSubjKey = filterMeta.key ? `filter-key-${filterMeta.key}` : '';
    const dataTestSubjValue = filterMeta.value;
    const dataTestSubjNegated = filterMeta.negate ? 'filter-negated' : '';
    const dataTestSubjDisabled = `filter-${filterMeta.disabled ? 'disabled' : 'enabled'}`;
    return `filter ${dataTestSubjDisabled} ${dataTestSubjKey} ${dataTestSubjValue} ${dataTestSubjNegated}`;
  }

  function getPanels() {
    const { negate, disabled } = filterMeta;
    return [
      {
        id: 0,
        items: [
          {
            name: i18n.translate('maps.filter.filterBar.editFilterButtonLabel', {
              defaultMessage: 'Edit filter',
            }),
            icon: 'pencil',
            panel: 1,
            'data-test-subj': 'editFilter',
          },
          {
            name: negate
              ? i18n.translate('maps.filter.filterBar.includeFilterButtonLabel', {
                  defaultMessage: 'Include results',
                })
              : i18n.translate('data.filter.filterBar.excludeFilterButtonLabel', {
                  defaultMessage: 'Exclude results',
                }),
            icon: negate ? 'plusInCircle' : 'minusInCircle',
            onClick: () => {
              setIsPopoverOpen(false);
              onToggleNegated();
            },
            'data-test-subj': 'negateFilter',
          },
          {
            name: disabled
              ? i18n.translate('data.filter.filterBar.enableFilterButtonLabel', {
                  defaultMessage: 'Re-enable',
                })
              : i18n.translate('data.filter.filterBar.disableFilterButtonLabel', {
                  defaultMessage: 'Temporarily disable',
                }),
            icon: `${disabled ? 'eye' : 'eyeClosed'}`,
            onClick: () => {
              setIsPopoverOpen(false);
              onToggleDisabled();
            },
            'data-test-subj': 'disableFilter',
          },
          {
            name: i18n.translate('data.filter.filterBar.deleteFilterButtonLabel', {
              defaultMessage: 'Delete',
            }),
            icon: 'trash',
            onClick: () => {
              setIsPopoverOpen(false);
              onRemove();
            },
            'data-test-subj': 'deleteFilter',
          },
        ],
      },
      {
        id: 1,
        width: 600,
        content: (
          <div>
            <FilterEditor
              content={JSON.stringify(filterMeta.params, null, 2)}
              label={filterMeta.alias!}
              onSubmit={onSubmit}
              onCancel={() => {
                setIsPopoverOpen(false);
              }}
            />
          </div>
        ),
      },
    ];
  }

  const badge = (
    <FilterView
      isDisabled={filterMeta.disabled}
      className={getClasses(filterMeta.negate, filterMeta.disabled)}
      onRemove={onRemove}
      valueLabel={filterMeta.alias || ''}
      onClick={handleClick}
      data-test-subj={getDataTestSubj()}
    />
  );

  return (
    <EuiPopover
      id={`popoverFor_filter${id}`}
      className={`globalFilterItem__popover`}
      anchorClassName={`globalFilterItem__popoverAnchor`}
      isOpen={isPopoverOpen}
      closePopover={() => {
        setIsPopoverOpen(false);
      }}
      button={badge}
      anchorPosition="downLeft"
      panelPaddingSize="none"
    >
      <EuiContextMenu initialPanelId={0} panels={getPanels()} />
    </EuiPopover>
  );
}
