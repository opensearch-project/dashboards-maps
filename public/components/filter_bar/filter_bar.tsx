/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
import classNames from 'classnames';
import React from 'react';

import { FilterItem } from './filter_item';
import { GeoShapeFilterMeta } from '../../../../../src/plugins/data/common';
import {
  disableGeoShapeFilterMeta,
  enableGeoShapeFilterMeta,
  toggleGeoShapeFilterMetaDisabled,
  toggleGeoShapeFilterMetaNegated,
} from './filter_actions';
import { FilterOptions } from './filter_options';

interface Props {
  filters: GeoShapeFilterMeta[];
  onFiltersUpdated?: (filters: GeoShapeFilterMeta[]) => void;
  className: string;
}
export const FilterBar = ({ filters, onFiltersUpdated, className }: Props) => {
  function updateFilters(items: GeoShapeFilterMeta[]) {
    if (onFiltersUpdated) {
      onFiltersUpdated(items);
    }
  }

  function renderItems(filterMeta: GeoShapeFilterMeta[]) {
    return filterMeta.map((meta, i) => (
      <EuiFlexItem key={i} grow={false} className="globalFilterBar__flexItem">
        <FilterItem
          id={`${i}`}
          filterMeta={meta}
          onUpdate={(newFilter) => onUpdate(i, newFilter)}
          onRemove={() => onRemove(i)}
        />
      </EuiFlexItem>
    ));
  }

  function onEnableAll() {
    const updatedFilters: GeoShapeFilterMeta[] = filters.map(enableGeoShapeFilterMeta);
    if (onFiltersUpdated) {
      onFiltersUpdated(updatedFilters);
    }
  }

  function onDisableAll() {
    const updatedFilters: GeoShapeFilterMeta[] = filters.map(disableGeoShapeFilterMeta);
    if (onFiltersUpdated) {
      onFiltersUpdated(updatedFilters);
    }
  }

  function onToggleAllNegated() {
    const updatedFilters: GeoShapeFilterMeta[] = filters.map(toggleGeoShapeFilterMetaNegated);
    if (onFiltersUpdated) {
      onFiltersUpdated(updatedFilters);
    }
  }

  function onToggleAllDisabled() {
    const updatedFilters: GeoShapeFilterMeta[] = filters.map(toggleGeoShapeFilterMetaDisabled);
    if (onFiltersUpdated) {
      onFiltersUpdated(updatedFilters);
    }
  }

  function onRemoveAll() {
    if (onFiltersUpdated) {
      onFiltersUpdated([]);
    }
  }

  function onRemove(i: number) {
    const updatedFilters = [...filters];
    updatedFilters.splice(i, 1);
    updateFilters(updatedFilters);
  }

  function onUpdate(i: number, filter: GeoShapeFilterMeta) {
    const current = [...filters];
    current[i] = filter;
    updateFilters(current);
  }

  const classes = classNames('globalFilterBar', className);

  return (
    <EuiFlexGroup
      className="globalFilterGroup"
      gutterSize="none"
      alignItems="flexStart"
      responsive={false}
    >
      <EuiFlexItem className="globalFilterGroup__branch" grow={false}>
        <FilterOptions
          onEnableAll={onEnableAll}
          onDisableAll={onDisableAll}
          onToggleAllNegated={onToggleAllNegated}
          onToggleAllDisabled={onToggleAllDisabled}
          onRemoveAll={onRemoveAll}
        />
      </EuiFlexItem>
      <EuiFlexItem className="globalFilterGroup__filterFlexItem">
        <EuiFlexGroup
          className={classes}
          wrap={true}
          responsive={false}
          gutterSize="xs"
          alignItems="center"
        >
          {renderItems(filters)}
        </EuiFlexGroup>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};
