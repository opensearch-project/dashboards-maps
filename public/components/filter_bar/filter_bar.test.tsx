/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { shallow } from 'enzyme';
import React from 'react';
import { FilterBar } from './filter_bar';
import { FILTERS, GeoShapeFilterMeta, ShapeFilter } from '../../../../../src/plugins/data/common';
import { GeoShapeRelation } from '@opensearch-project/opensearch/api/types';
import { Polygon } from 'geojson';

it('renders one filter inside filter bar', () => {
  const mockCallback = jest.fn();
  const mockPolygon: Polygon = {
    type: 'Polygon',
    coordinates: [
      [
        [74.006, 40.7128],
        [71.0589, 42.3601],
        [73.7562, 42.6526],
        [74.006, 40.7128],
      ],
    ],
  };
  const expectedParams: {
    shape: ShapeFilter;
    relation: GeoShapeRelation;
  } = {
    shape: mockPolygon,
    relation: 'intersects',
  };
  const mockFilterMeta: GeoShapeFilterMeta = {
    alias: 'mylabel',
    disabled: false,
    negate: false,
    type: FILTERS.GEO_SHAPE,
    params: expectedParams,
  };
  const filterBar = shallow(
    <FilterBar
      filters={[mockFilterMeta]}
      onFiltersUpdated={mockCallback}
      className={'some-class-name'}
    />
  );
  expect(filterBar).toMatchSnapshot();
});
