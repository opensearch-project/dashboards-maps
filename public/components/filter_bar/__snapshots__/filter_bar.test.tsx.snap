// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders one filter inside filter bar 1`] = `
<EuiFlexGroup
  alignItems="flexStart"
  className="globalFilterGroup"
  gutterSize="none"
  responsive={false}
>
  <EuiFlexItem
    className="globalFilterGroup__branch"
    grow={false}
  >
    <FilterOptions
      onDisableAll={[Function]}
      onEnableAll={[Function]}
      onRemoveAll={[Function]}
      onToggleAllDisabled={[Function]}
      onToggleAllNegated={[Function]}
    />
  </EuiFlexItem>
  <EuiFlexItem
    className="globalFilterGroup__filterFlexItem"
  >
    <EuiFlexGroup
      alignItems="center"
      className="globalFilterBar some-class-name"
      gutterSize="xs"
      responsive={false}
      wrap={true}
    >
      <EuiFlexItem
        className="globalFilterBar__flexItem"
        grow={false}
        key="0"
      >
        <FilterItem
          filterMeta={
            Object {
              "alias": "mylabel",
              "disabled": false,
              "negate": false,
              "params": Object {
                "relation": "intersects",
                "shape": Object {
                  "coordinates": Array [
                    Array [
                      Array [
                        74.006,
                        40.7128,
                      ],
                      Array [
                        71.0589,
                        42.3601,
                      ],
                      Array [
                        73.7562,
                        42.6526,
                      ],
                      Array [
                        74.006,
                        40.7128,
                      ],
                    ],
                  ],
                  "type": "Polygon",
                },
              },
              "type": "geo_shape",
            }
          }
          id="0"
          onRemove={[Function]}
          onUpdate={[Function]}
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiFlexItem>
</EuiFlexGroup>
`;
